File: 11-error-handling/src/main.c
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <string.h>

int read_positive_integer(void) {
    int rc = 0;
    int buf_size = 16;
    char *buf = malloc(buf_size);

    printf("Enter a number: ");

    if (fgets(buf, buf_size, stdin) == NULL) {
	printf("No input!\n");
	rc = -EINVAL;
	goto clean;
    }

    buf[strcspn(buf, "\n")] = '\0';

    errno = 0;
    rc = strtol(buf, NULL, 10);

    if (errno != 0) {
	rc = -EINVAL;
	goto clean;
    }
    
 clean:
    free(buf);

    return rc;
}

int main(void) {
    int number = read_positive_integer();

    if (number < 0) {
	printf("Error: %i\n", number);
    } else {
	printf("Read: '%i'\n", number);
    }
}

File: 11-error-handling/src/main.rs
use std::{io::Write, num::ParseIntError};

#[derive(Debug)]
enum Error {
    Io(std::io::Error),
    Parse(ParseIntError),
}

impl From<std::io::Error> for Error {
    fn from(value: std::io::Error) -> Self {
	Self::Io(value)
    }
}

impl From<ParseIntError> for Error {
    fn from(value: ParseIntError) -> Self {
	Self::Parse(value)
    }
}

fn read_positive_integer() -> Result<u32, Error> {
    let stdin = std::io::stdin();
    let mut buf = String::new();

    print!("Enter a number: ");
    std::io::stdout().flush()?;

    stdin.read_line(&mut buf)?;

    let number: u32 = buf.trim().parse()?;

    Ok(number)
}

fn main() {
    match read_positive_integer() {
        Ok(number) => println!("Read: {}", number),
        Err(err) => println!("Error: {:?}", err),
    }
}
